### PR TITLE
Dataset loading

### PR DESCRIPTION
**Pull request by rstevanak**

### CODE REVIEW

**Comment by mrshu** [comment]
@rstevanak before we get to the actual code review, please fix the code formatting issues outlined here: https://travis-ci.org/NaiveNeuron/PKSpace/builds/198847983#L179-L192

Thanks

File: pkspace/utils/loaders.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class Loader:
+    def load(self, path, split):
+        pass
+
+
+class PKspaceLoader(Loader):
+    def __get_spaces__(self, src, desc):
+        finsize = (80, 80)
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak This should probably be provided to the function as a keyword parameter.

File: pkspace/utils/loaders.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import numpy as np
+
+
+class Loader:
+    def load(self, path, split):
+        pass
+
+
+class PKspaceLoader(Loader):
+    def __get_spaces__(self, src, desc):
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak why does this function has so many underscores in it? As far as I see, this makes it  a [Special Name](https://docs.python.org/3.4/reference/datamodel.html#specialnames), which is intended for Python's internal use. If we are afraid that this function may be overridden by something that subclasses `PKspaceLoader`, it is probably better to call it `_get_spaces`.

See [this](https://shahriar.svbtle.com/underscores-in-python) for more details.

File: pkspace/utils/loaders.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import cv2
+import numpy as np
+
+
+class Loader:
+    def load(self, path, split):
+        pass
+
+
+class PKspaceLoader(Loader):
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak since we call this project `PKSpace` we may as well call this class `PKSpaceLoader`.

File: pkspace/utils/loaders.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    def __get_spaces__(self, src, desc):
+        finsize = (80, 80)
+        spaces = []
+        answers = []
+        for space in desc.get('spots'):
+            points = space.get('points')
+            angle = space.get('rotation')
+            nove_pole = np.hstack(
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak please let's either use English variable names, or some other language, but let us also make sure we do not mingle two of them together.

File: pkspace/utils/loaders.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                (np.asarray(points), np.ones((len(points), 1))))
+            center = np.mean(nove_pole, 0)[:2]
+            center = (center[0], center[1])
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak can be also rewritten as `center = *center`, but this is probably much more clear.

File: pkspace/utils/loaders.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            rows, cols, tmp = src.shape
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak if you do not intend to use the third value, please use `_` there instead of a `tmp` variable which needs to be initialized this way.

File: pkspace/utils/loaders.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            m = cv2.getRotationMatrix2D(center, -angle, 1)
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak In order to make the code more readable, I suggest you use whitespace to separate different logical blocks. This may be a good line to do that.

File: pkspace/utils/loaders.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    def load(self, path, split):
+        pass
+
+
+class PKspaceLoader(Loader):
+    def __get_spaces__(self, src, desc):
+        finsize = (80, 80)
+        spaces = []
+        answers = []
+        for space in desc.get('spots'):
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak I can see why you may want to do `desc.get('sports')` here -- if the `sports` key does not exist in `desc` you would get `None` in return and the whole loop won't get executed. 

I do not really see why would you do the same on the lines below: if there is no `points` for a specific space, your `points` variable will be set to `None` and the code below that makes use of it will play with it just as if nothing happened. it will of course produce wrong results, but you'll notice that much later. I believe it would be better to let it trigger `KeyError` or do not let the code below execute if either `points` or `angle` is `None`. in case of `angle`, you may want to define some default, but I do not see what else could be done for `points`.

File: pkspace/utils/loaders.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            nove_pole[:, 1] = np.clip(nove_pole[:, 1], 0, rows)
+            minima = nove_pole.min(axis=0)
+            maxima = nove_pole.max(axis=0)
+            min_x, min_y = int(minima[0]), int(minima[1])
+            max_x, max_y = int(maxima[0]), int(maxima[1])
+            rotated = cv2.warpAffine(src, m, (cols, rows))
+            roi = rotated[min_y:max_y, min_x:max_x]
+            fin = cv2.resize(roi, finsize).flatten()
+            spaces.append(fin)
+            answers.append(space.get('occupied', -1))
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak at this point it does make sense to do `space.get`. I am wondering, however, if you could make it more clear that this is the actual label from the dataset.

File: pkspace/utils/loaders.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        return spaces, answers
+
+    def load(self, path, split, extension='.png'):
+        all_parking_spaces = []
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak what was previously called `spaces` is now called `parking_spaces` -- let's use just one of them for consistency.

File: pkspace/utils/loaders.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        all_answers = []
+        extension = "*{}".format(extension)
+        for file in glob.glob(os.path.join(path, extension)):
+            img = cv2.imread(file)
+            jsonfile = open("{}.json".format(os.path.splitext(file)[0]))
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak please extract `os.path.splitext(file)[0]` into a variable above, so that it is obvious that it is the json filename.

File: pkspace/utils/loaders.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+            desc = json.load(jsonfile)
+            spaces, answers = self.__get_spaces__(img, desc)
+            all_parking_spaces.append(spaces)
+            all_answers.append(answers)
+        all_parking_spaces = np.asarray(all_parking_spaces)
+        all_answers = np.asarray(all_answers)
+        assert split < 1, 'split needs to be smaller than 1'
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak once again, use whitespace to separate logical blocks, especially when there is a new keyword around the block.

File: pkspace/utils/loaders.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        x_train = all_parking_spaces[indices[:train_size]]
+        y_train = all_answers[indices[:train_size]]
+        x_test = all_parking_spaces[indices[:-train_size]]
+        y_test = all_answers[indices[:-train_size]]
+        return x_train, x_test, y_train, y_test
+
+
+class PKlotLoader(Loader):
+    def __get_spaces__(self, src, tabroot):
+        fin_size = (80, 80)
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak same as above

File: pkspace/utils/loaders.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+            min_x, min_y = int(minima[0]), int(minima[1])
+            max_x, max_y = int(maxima[0]), int(maxima[1])
+            rotated = cv2.warpAffine(src, m, (cols, rows))
+            roi = rotated[min_y:max_y, min_x:max_x]
+            fin = cv2.resize(roi, finsize).flatten()
+            spaces.append(fin)
+            answers.append(space.get('occupied', -1))
+        return spaces, answers
+
+    def load(self, path, split, extension='.png'):
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak seems like this function may benefit from getting a keyword parameter `size` which would specify the size of "patches" it generates.

File: pkspace/utils/loaders.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+class PKlotLoader(Loader):
+    def __get_spaces__(self, src, tabroot):
+        fin_size = (80, 80)
+        spaces = []
+        answers = []
+        for space in tabroot:
+            center = int(space[0][0].get('x')), int(space[0][0].get('y'))
+            h, w = int(space[0][1].get('h')), int(space[0][1].get('w'))
+            m = cv2.getRotationMatrix2D(center, int(space[0][2].get('d')), 1)
+            rows, cols, tmp = src.shape
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak same as above.

File: pkspace/utils/loaders.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        return x_train, x_test, y_train, y_test
+
+
+class PKlotLoader(Loader):
+    def __get_spaces__(self, src, tabroot):
+        fin_size = (80, 80)
+        spaces = []
+        answers = []
+        for space in tabroot:
+            center = int(space[0][0].get('x')), int(space[0][0].get('y'))
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak it really does not make sense to use `.get` here as it is longer and does not help anything -- if the result of `space[0][0].get('x')` is `None`, then what follows is `int(None)` which trows an exception.

Please use `space[0][0]['x']` instead.

File: pkspace/utils/loaders.py (hunk 9 above)
Comment (mrshu) [reply pkspace/utils/loaders.py]: Moreover, what does the value of `-1` mean?

File: pkspace/utils/loaders.py [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+            roi = dst[
+                int(center[1] - h / 2):int(center[1] + h / 2),
+                int(center[0] - w / 2):int(center[0] + w / 2)]
+            fin = cv2.resize(roi, fin_size).flatten()
+            answers.append(space.get('occupied', 1))
+            spaces.append(fin)
+        return spaces, answers
+
+    def load(self, path, split, extension='.jpg'):
+        all_parking_spaces = []
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak same as above.

File: pkspace/utils/loaders.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+            dst = cv2.warpAffine(src, m, (cols, rows))
+            roi = dst[
+                int(center[1] - h / 2):int(center[1] + h / 2),
+                int(center[0] - w / 2):int(center[0] + w / 2)]
+            fin = cv2.resize(roi, fin_size).flatten()
+            answers.append(space.get('occupied', 1))
+            spaces.append(fin)
+        return spaces, answers
+
+    def load(self, path, split, extension='.jpg'):
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak same as above, the `size` keyword may be useful here.

File: pkspace/utils/loaders.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+        all_parking_spaces = []
+        all_answers = []
+        extension = "*{}".format(extension)
+        for file in glob.glob(os.path.join(path, extension)):
+            img = cv2.imread(file)
+            e = xml.etree.ElementTree.parse(
+                "{}.xml".format(os.path.splitext(file)[0]))
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak if you move `os.path.splitext(file)[0]` into its own variable, this block of code will look much nicer.

File: pkspace/utils/loaders.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        indices = np.random.permutation(len(all_parking_spaces))
+        x_train = all_parking_spaces[indices[:train_size]]
+        y_train = all_answers[indices[:train_size]]
+        x_test = all_parking_spaces[indices[:-train_size]]
+        y_test = all_answers[indices[:-train_size]]
+        return x_train, x_test, y_train, y_test
+
+
+class PKlotLoader(Loader):
+    def __get_spaces__(self, src, tabroot):
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak same as above with regards to the underscores.

File: pkspace/utils/loaders.py [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        train_size = int(np.ceil(len(all_parking_spaces) * split))
+        indices = np.random.permutation(len(all_parking_spaces))
+        x_train = all_parking_spaces[indices[:train_size]]
+        y_train = all_answers[indices[:train_size]]
+        x_test = all_parking_spaces[indices[:-train_size]]
+        y_test = all_answers[indices[:-train_size]]
+        return x_train, x_test, y_train, y_test
+
+
+class PKlotLoader(Loader):
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak same as above, this should probably be `PKLotLoader`.

File: pkspace/utils/loaders.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+            img = cv2.imread(file)
+            e = xml.etree.ElementTree.parse(
+                "{}.xml".format(os.path.splitext(file)[0]))
+            root = e.getroot()
+            spaces, answers = self.__get_spaces__(img, root)
+            all_parking_spaces.append(spaces)
+            all_answers.append(answers)
+        all_parking_spaces = np.asarray(all_parking_spaces)
+        all_answers = np.asarray(all_answers)
+        assert split < 1, 'split needs to be smaller than 1'
```
Comment (mrshu) [inline pkspace/utils/loaders.py]: @rstevanak same as above.

File: wsgi.py [hunk 23]
```diff
@@ -1,18 +1,18 @@
...
-    execfile(virtualenv, dict(__file__=virtualenv))
+    execfile(virtualenv, dict(__file__=virtualenv))  # noqa
 except IOError:
     pass
 #
 # IMPORTANT: Put any additional includes below this line.  If placed above this
 # line, it's possible required libraries won't be in your searchable path
 #
 
-from configapp.configapp import app as application
```
Comment (mrshu) [inline wsgi.py]: @rstevanak why was this removed? It is used at [line 22](https://github.com/NaiveNeuron/PKSpace/pull/16/files#diff-ea8aaf2b76cffeafd45054ae9c1ff347R22) -- the whole thing will fail without it.

**Review by mrshu — CHANGES_REQUESTED** [review]
All in all, I would say that this is a good start -- I did not run into anything with major issues. Most of the issues I found were nitpicks or things that did not make sense to me.

File: pkspace/utils/loaders.py (hunk 6 above)
Comment (vidriduch) [reply pkspace/utils/loaders.py]: @rstevanak or maybe just use ``rows, cols = src.shape[:2]``

File: pkspace/utils/loaders.py [hunk 24]
```diff
@@ -21,40 +21,17 @@ def __get_spaces__(self, src, desc):
...
-            # min_x, min_y = minima[:2]
-            # max_x, max_y = maxima[:2]
-
+            nove_pole = m.dot(nove_pole.T).T
+            nove_pole[:, 0] = np.clip(nove_pole[:, 0], 0, cols)
+            nove_pole[:, 1] = np.clip(nove_pole[:, 1], 0, rows)
+            minima = nove_pole.min(axis=0)
+            maxima = nove_pole.max(axis=0)
+            min_x, min_y = int(minima[0]), int(minima[1])
+            max_x, max_y = int(maxima[0]), int(maxima[1])
```
Comment (vidriduch) [inline pkspace/utils/loaders.py]: @rstevanak I belive this can be reduce to
```python
min_x, min_y = nove_pole.min(axis=0).astype('uint32')
max_x, max_y = nove_pole.max(axis=0).astype('uint32')
```
which I personally think easier expresses what we are trying to do there.

File: pkspace/utils/loaders.py (hunk 14 above)
Comment (rstevanak) [reply pkspace/utils/loaders.py]: Do you mean size to be finsize in __get_spaces__?

File: pkspace/utils/loaders.py (hunk 9 above)
Comment (rstevanak) [reply pkspace/utils/loaders.py]: -1 was supposed to be my internal code for unmarked space, but now it is obsolete. Can I expect dataset to be always fully labeled so I can get rid of it?

File: pkspace/utils/loaders.py (hunk 24 above)
Comment (mrshu) [reply pkspace/utils/loaders.py]: Thanks @Foxiy, I agree.

**Comment by mrshu** [comment]
This is just utterly broken.
